infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/provision/AsyncStorageProvisioner.java

```diff
@@ -0,0 +1,370 @@
+    if (!isPodExist) {
+      Pod pod = createStoragePod(namespace, configMapName);
+      oc.pods().inNamespace(namespace).create(pod);
+    }
+
+    boolean isServiceExist =
+        oc.services()
+            .inNamespace(namespace)
+            .list()
+            .getItems()
+            .stream()
+            .anyMatch(
+                (Predicate<Service>)
+                    service -> service.getMetadata().getName().equals(ASYNC_STORAGE));
+
+    if (!isServiceExist) {
+      Service service = createStorageService(namespace);
+      oc.services().inNamespace(namespace).create(service);
+    }
+  }
+
+  /** Get or create new pair of SSH keys, this is need for securing rsync connection */
+  private List<SshPairImpl> getOrCreateSshPairs(
+      RuntimeIdentity identity, OpenShiftEnvironment osEnv) throws InfrastructureException {
+    List<SshPairImpl> sshPairs;
+    try {
+      sshPairs = sshManager.getPairs(identity.getOwnerId(), "internal");
+    } catch (ServerException e) {
+      String message = format("Unable to get SSH Keys. Cause: %s", e.getMessage());
+      LOG.warn(message);
+      osEnv.addWarning(
+          new WarningImpl(
+              NOT_ABLE_TO_PROVISION_SSH_KEYS,
+              format(NOT_ABLE_TO_PROVISION_SSH_KEYS_MESSAGE, message)));
+      throw new InfrastructureException(e);
+    }
+    if (sshPairs.isEmpty()) {
+      try {
+        sshPairs =
+            singletonList(sshManager.generatePair(identity.getOwnerId(), "internal", SSH_KEY_NAME));
+      } catch (ServerException | ConflictException e) {
+        String message =
+            format(
+                "Unable to generate the SSH key for async storage service. Cause: %S",
+                e.getMessage());
+        LOG.warn(message);
+        osEnv.addWarning(
+            new WarningImpl(
+                NOT_ABLE_TO_PROVISION_SSH_KEYS,
+                format(NOT_ABLE_TO_PROVISION_SSH_KEYS_MESSAGE, message)));
+        throw new InfrastructureException(e);
+      }
+    }
+    return sshPairs;
+  }
+
+  /** Create configmap with public part of SSH key */
+  private ConfigMap createConfigMap(
+      String namespace, String configMapName, RuntimeIdentity identity, OpenShiftEnvironment osEnv)
+      throws InfrastructureException {
+    List<SshPairImpl> sshPairs = getOrCreateSshPairs(identity, osEnv);
+    if (sshPairs == null) {
+      return null;
+    }
+    SshPair sshPair = sshPairs.get(0);
+    Map<String, String> sshConfigData =
+        ImmutableMap.of(AUTHORIZED_KEYS, sshPair.getPublicKey() + "\n");
+    return new ConfigMapBuilder()
+        .withNewMetadata()
+        .withName(configMapName)
+        .withNamespace(namespace)
+        .endMetadata()
+        .withData(sshConfigData)
+        .build();
+  }
+
+  /**
+   * Create storage Pod with container with mounted volume for storing project source backups, SSH
+   * key and exposed port for rsync connection
+   */
+  private Pod createStoragePod(String namespace, String configMap) {
```

Comment:
might be good to rename to createAsyncStoragePod

Reply:
done